src/basic/process-util.c

```diff
@@ -699,6 +699,54 @@ int wait_for_terminate_and_warn(const char *name, pid_t pid, bool check_exit_cod
         return -EPROTO;
 }
 
+int wait_for_terminate_with_timeout(pid_t pid, usec_t timeout) {
+        sigset_t mask;
+        int r;
+        siginfo_t status;
+        usec_t until;
+
+        assert_se(sigemptyset(&mask) == 0);
+        assert_se(sigaddset(&mask, SIGCHLD) == 0);
+
+        /* Drop into a sigtimewait-based timeout. Waiting for the
+         * remount pid to exit. */
+        until = now(CLOCK_MONOTONIC) + timeout;
+        for (;;) {
+                usec_t n;
+                struct timespec ts;
+
+                n = now(CLOCK_MONOTONIC);
+                if (n >= until)
+                        break;
+
+                zero(status);
+
+                r = sigtimedwait(&mask, NULL, timespec_store(&ts, until - n));
+                /* Assuming we woke due to the child exiting. */
+                if (waitid(P_PID, pid, &status, WEXITED|WNOHANG) == 0) {
+                        if (status.si_pid == pid)
+                                /* This is the correct child */
+                                return (status.si_status == EXIT_SUCCESS) ? 0 : -EPROTO;
```

Comment:
if status.si_code is not CLD_EXITED then we should also return -EPROTO here. After all theoretically our child process might also die due to signal, not just with a clean exit code